plugins/modem-manager/fu-sahara-loader.c

```diff
@@ -0,0 +1,562 @@
+{
+	g_return_val_if_fail(self != NULL, FALSE);
+
+	return fu_usb_device_is_open(self->usb_device);
+}
+
+GByteArray *
+fu_sahara_loader_qdl_read(FuSaharaLoader *self, GError **error)
+{
+	gsize actual_len = 0;
+	g_autoptr(GByteArray) buf = g_byte_array_sized_new(SAHARA_RAW_BUFFER_SIZE);
+	fu_byte_array_set_size(buf, SAHARA_RAW_BUFFER_SIZE);
+
+	if (!g_usb_device_bulk_transfer(fu_usb_device_get_dev(self->usb_device),
+					self->ep_in,
+					buf->data,
+					buf->len,
+					&actual_len,
+					IO_TIMEOUT_MS,
+					NULL,
+					error)) {
+		g_prefix_error(error, "failed to do bulk transfer (read): ");
+		return NULL;
+	}
+
+	g_byte_array_set_size(buf, actual_len);
+
+	g_debug("received %" G_GSIZE_FORMAT " bytes", actual_len);
+
+	return g_steal_pointer(&buf);
+}
+
+gboolean
+fu_sahara_loader_qdl_write(FuSaharaLoader *self, const guint8 *data, gsize sz, GError **error)
+{
+	FuChunk *chk;
+	gsize total_sent = 0;
+	gsize actual_len = 0;
+	g_autoptr(GPtrArray) chunks = NULL;
+	g_autoptr(GByteArray) bytes = NULL;
+
+	if (g_getenv("FWUPD_MODEM_MANAGER_VERBOSE") != NULL)
+		g_debug("writing %" G_GSIZE_FORMAT " bytes", sz);
+
+	/* copy const data to mutable GByteArray */
+	bytes = g_byte_array_sized_new(sz);
+	bytes = g_byte_array_append(bytes, data, sz);
+	chunks = fu_chunk_array_mutable_new(bytes->data, bytes->len, 0, 0, self->maxpktsize_out);
+	for (guint i = 0; i < chunks->len; i++) {
+		chk = g_ptr_array_index(chunks, i);
+
+		if (g_getenv("FWUPD_MODEM_MANAGER_VERBOSE") != NULL)
+			g_debug("sending %u bytes", fu_chunk_get_data_sz(chk));
+
+		if (!g_usb_device_bulk_transfer(fu_usb_device_get_dev(self->usb_device),
+						self->ep_out,
+						fu_chunk_get_data_out(chk),
+						fu_chunk_get_data_sz(chk),
+						&actual_len,
+						IO_TIMEOUT_MS,
+						NULL,
+						error)) {
+			g_prefix_error(error, "failed to do bulk transfer (write): ");
+			return FALSE;
+		}
+		if (actual_len != fu_chunk_get_data_sz(chk)) {
+			g_set_error(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_INVALID_DATA,
+				    "only wrote %" G_GSIZE_FORMAT "bytes",
+				    actual_len);
+			return FALSE;
+		}
+		total_sent += actual_len;
+	}
+	if (sz % self->maxpktsize_out == 0) {
+		if (!g_usb_device_bulk_transfer(fu_usb_device_get_dev(self->usb_device),
+						self->ep_out,
+						NULL,
+						0,
+						NULL,
+						IO_TIMEOUT_MS,
+						NULL,
+						error)) {
+			g_prefix_error(error, "failed to do bulk transfer (write): ");
+			return FALSE;
+		}
+		if (g_getenv("FWUPD_MODEM_MANAGER_VERBOSE") != NULL)
+			g_debug("sent zlp");
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_sahara_loader_write_prog(FuSaharaLoader *self,
+			    guint32 offset,
+			    guint32 length,
+			    GBytes *prog,
+			    GError **error)
+{
+	gsize sz;
+	const guint8 *data = g_bytes_get_data(prog, &sz);
+
+	g_return_val_if_fail(offset + length <= sz, FALSE);
+
+	g_debug("SENDING --> RAW_DATA: %u bytes (offset = %u, total len = %" G_GSIZE_FORMAT ")",
+		length,
+		offset,
+		sz);
```

Comment:
if you have to do this, can you put it behind FWUPD_MODEM_MANAGER_VERBOSE please